validate.go

```diff
@@ -18,34 +37,83 @@ func validate(payload []byte) ([]byte, error) {
+		"request.object.spec.securityContext.sysctls")
 
 	if !data.Exists() {
-		logger.Warn("cannot read object name from metadata: accepting request")
+		logger.Warn("pod doesn't specify sysctls: accepting request")
 		return kubewarden.AcceptRequest()
 	}
-	name := data.String()
 
-	logger.DebugWithFields("validating ingress object", func(e onelog.Entry) {
-		namespace := gjson.GetBytes(payload, "request.object.metadata.namespace").String()
+	logger.DebugWithFields("validating pod object", func(e onelog.Entry) {
+		name := gjson.GetBytes(payload, "request.object.metadata.name").String()
+		namespace := gjson.GetBytes(payload,
+			"request.object.metadata.namespace").String()
 		e.String("name", name)
 		e.String("namespace", namespace)
 	})
 
-	if settings.DeniedNames.Contains(name) {
-		logger.InfoWithFields("rejecting ingress object", func(e onelog.Entry) {
+	// build set of safe sysctls:
+	knownSafeSysctlsArray := getSafeSysctls()
+	knownSafeSysctls := mapset.NewThreadUnsafeSet()
+	for i := 0; i < len(knownSafeSysctlsArray); i++ {
+		knownSafeSysctls.Add(knownSafeSysctlsArray[i])
+	}
+
+	// build set of prefixes from patterns of forbidden sysctls:
+	globForbiddenSysctls := mapset.NewThreadUnsafeSet()
+	for _, elem := range settings.ForbiddenSysctls.ToSlice() {
+		if strings.HasSuffix(elem.(string), "*") {
+			globForbiddenSysctls.Add(strings.TrimSuffix(elem.(string), "*"))
+		}
+	}
+
+	data.ForEach(func(key, value gjson.Result) bool {
+		sysctl := gjson.Get(value.String(), "name").String()
+
+		if settings.ForbiddenSysctls.Contains(sysctl) {
+			err = fmt.Errorf("sysctl %s is on the forbidden list", sysctl)
+			return false // stop iterating
+		}
+
+		// if sysctl matches a pattern, it is forbidden:
+		for _, elem := range globForbiddenSysctls.ToSlice() {
+			if strings.HasPrefix(sysctl, elem.(string)) {
+				err = fmt.Errorf("sysctl %s is on the forbidden list", sysctl)
+				return false // stop iterating
+			}
+		}
+
+		// if sysctl is not on the safe list nor an exception, it is forbidden:
+		if !knownSafeSysctls.Contains(sysctl) &&
+			!settings.AllowedUnsafeSysctls.Contains(sysctl) {
+			err = fmt.Errorf("sysctl %s is not on safe list, nor is in the allowedUnsafeSysctls list",
+				sysctl)
+			return false // stop iterating
+		}
+		return true // continue iterating
+	})
+
+	if err != nil {
+		logger.DebugWithFields("rejecting pod object", func(e onelog.Entry) {
+			name := gjson.GetBytes(payload, "request.object.metadata.name").String()
+			namespace := gjson.GetBytes(payload, "request.object.metadata.namespace").String()
 			e.String("name", name)
-			e.String("denied_names", settings.DeniedNames.String())
+			e.String("namespace", namespace)
 		})
 
 		return kubewarden.RejectRequest(
-			kubewarden.Message(
-				fmt.Sprintf("The '%s' name is on the deny list", name)),
+			kubewarden.Message(err.Error()),
 			kubewarden.NoCode)
 	}
 
-	logger.Info("accepting ingress object")
+	logger.Info("accepting pod object")
```

Comment:
I would drop that. Keep in mind, the go template added logging a bit too much for my taste, but just to give an idea about how to use it. Maybe we should cleanup the template too

Reply:
Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/74ce514b0e87f365c2473773e7d04b22044d254e.
Opened https://github.com/kubewarden/go-policy-template/pull/8.